### PR TITLE
Fix FileResponse sending empty chunked body on 304

### LINKED ISSUE #2143: FileResponse sends empty chunked body when returning 304
## Long story short

`FileResponse` (and by extension, `add_static`) may return `Transfer-Encoding: chunked` and the empty chunked body when returning a 304 Not Modified, in violation of [RFC 2616 §10.3.5](https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.3.5), which states: "The 304 response MUST NOT contain a message-body, and thus is always terminated by the first empty line after the header fields."

## Expected behaviour

A 304 response with no body.

## Actual behaviour

A 304 response with `Transfer-Encoding: chunked` and a 1-byte body (`0`).

## Steps to reproduce
1. Run this with an empty `index.html` in the same directory:
```
from aiohttp import web
from pathlib import Path

app = web.Application()
app.router.add_static('/', Path(__file__).parent)
web.run_app(app, port=8080)
```

2. Send an HTTP/1.1 request with an `If-Modified-Since` that triggers a 304 Not Modified, e.g.
```
# nc localhost 8080
GET /index.html HTTP/1.1
If-Modified-Since: Fri, 28 Jul 2017 18:42:00 GMT

```

The response will be something like:

```
HTTP/1.1 304 Not Modified
Transfer-Encoding: chunked
Content-Type: application/octet-stream
Date: Fri, 28 Jul 2017 18:42:34 GMT
Server: Python/3.6 aiohttp/2.2.3

0

```

## Your environment

This is using aiohttp 2.2.3, the newest version at this time.

### PR DESCRIPTION
**Pull request by es1024**
## What do these changes do?

Fixes a bug where FileResponse/add_static will include a non-empty body and use Transfer-Encoding: chunked for 304 responses.

## Are there changes in behavior for the user?

FileResponse/add_static now won't give malformed 304 responses with bodies.

## Related issue number

This fixes #2143.

## Checklist

- [X] I think the code is well written
- [X] Unit tests for the changes exist
- [n/a] Documentation reflects the changes
- [X] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names. 
- [X] Add a new news fragment into the `changes` folder
  * name it `<issue_id>.<type>` for example (588.bug)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

### CODE REVIEW

File: tests/test_web_sendfile_functional.py [hunk 1]
```diff
@@ -165,7 +165,9 @@ def handler(request):
     resp.close()
 
     resp = yield from client.get('/', headers={'If-Modified-Since': lastmod})
+    body = yield from resp.read()
     assert 304 == resp.status
+    assert b'' == body
```
Comment (cecton) [inline tests/test_web_sendfile_functional.py]: Hello @es1024 ,

According to what I read on the RFC:
 *  http://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.3
 *  http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.3.5

The Content-Length header should not even be there. Can you add an assert to enforce it?

File: aiohttp/web_response.py [hunk 2]
```diff
@@ -362,7 +362,8 @@ def _start(self, request,
                 del headers[CONTENT_LENGTH]
         elif self._length_check:
             writer.length = self.content_length
-            if writer.length is None and version >= HttpVersion11:
+            if (writer.length is None and version >= HttpVersion11 and
```
Comment (asvetlov) [inline aiohttp/web_response.py]: I think fix should be applied to `web_fileresponse.py` instead of base `StreamResponse` class.

**Comment by asvetlov** [comment]
pretty sure `BIN  aiohttp/.web_response.py.swo` was not intentionally added.

File: aiohttp/web_fileresponse.py [hunk 3]
```diff
@@ -179,6 +179,7 @@ def prepare(self, request):
         modsince = request.if_modified_since
         if modsince is not None and st.st_mtime <= modsince.timestamp():
             self.set_status(HTTPNotModified.status_code)
+            self._length_check = False
```
Comment (asvetlov) [inline aiohttp/web_fileresponse.py:182]: This line looks hacky but I could live with it.
Looks like `StreamResponse` is too coupled with `Response` class, let's left it for other PR

File: tests/test_web_sendfile_functional.py (hunk 1 above)
Comment (cecton) [reply tests/test_web_sendfile_functional.py]: Thx!

This habit of so many dev to always use `.get()` 😕 This is so much more readable (and slightly more correct):

```python
assert "Content-Length" not in resp.headers
```

(Don't bother though, not important)

**Comment by cecton** [comment]
@asvetlov can I do the merge? (I will drop the bin don't worry)

**Comment by asvetlov** [comment]
@cecton please do

**Comment by cecton** [comment]
Fixed by cc614ca3ecbcb3d49854a2a7411cf64eb724923d

**Pull request closed without merging**